src/cargo/sources/git/utils.rs

```diff
@@ -557,29 +558,158 @@ fn with_authentication<T, F>(url: &str, cfg: &git2::Config, mut f: F)
 
         // Create a local anonymous remote in the repository to fetch the url
-        let mut remote = repo.remote_anonymous(url)?;
+        let mut remote = repo.remote_anonymous(url.as_str())?;
         let mut opts = git2::FetchOptions::new();
         opts.remote_callbacks(cb)
             .download_tags(git2::AutotagOption::All);
 
         network::with_retry(config, || {
+            debug!("initiating fetch of {} from {}", refspec, url);
             remote.fetch(&[refspec], Some(&mut opts), None)
                 .map_err(CargoError::from)
         })?;
         Ok(())
     })
 }
+
+/// Cargo has a bunch of long-lived git repositories in its global cache and
+/// some, like the index, are updated very frequently. Right now each update
+/// creates a new "pack file" inside the git database, and over time this can
+/// cause bad performance and bad current behavior in libgit2.
+///
+/// One pathological use case today is where libgit2 opens hundreds of file
+/// descriptors, getting us dangerously close to blowing out the OS limits of
+/// how many fds we can have open. This is detailed in #4403.
+///
+/// To try to combat this problem we attempt a `git gc` here. Note, though, that
+/// we may not even have `git` installed on the system! As a result we
+/// opportunistically try a `git gc` when the pack directory looks too big, and
+/// failing that we just blow away the repository and start over.
+fn maybe_gc_repo(repo: &mut git2::Repository) -> CargoResult<()> {
+    // Here we arbitrarily declare that if you have more than 100 files in your
+    // `pack` folder that we need to do a gc.
+    let entries = match repo.path().join("objects/pack").read_dir() {
+        Ok(e) => e.count(),
+        Err(_) => {
+            debug!("skipping gc as pack dir appears gone");
+            return Ok(())
+        }
+    };
+    if entries < 100 {
+        debug!("skipping gc as there's only {} pack files", entries);
+        return Ok(())
+    }
+
+    // First up, try a literal `git gc` by shelling out to git. This is pretty
+    // likely to fail though as we may not have `git` installed. Note that
+    // libgit2 doesn't currently implement the gc operation, so there's no
+    // equivalent there.
+    match Command::new("git").arg("gc").current_dir(repo.path()).output() {
+        Ok(out) => {
+            debug!("git-gc status: {}\n\nstdout ---\n{}\nstderr ---\n{}",
+                   out.status,
+                   String::from_utf8_lossy(&out.stdout),
+                   String::from_utf8_lossy(&out.stderr));
+            if out.status.success() {
+                let new = git2::Repository::open(repo.path())?;
+                mem::replace(repo, new);
+                return Ok(())
+            }
+        }
+        Err(e) => debug!("git-gc failed to spawn: {}", e),
+    }
+
+    // Alright all else failed, let's start over
+    if repo.is_bare() {
+        let path = repo.path().to_path_buf();
+        drop(fs::remove_dir_all(&path));
```

Comment:
Hm, are there any race conditions here? Would this fast path work: https://github.com/rust-lang/cargo/blob/053b634cd2c85a18b7b2568ef0056d649112a9e0/src/cargo/sources/registry/remote.rs#L58 ?

Reply:
Yeah that's a good point, I'll investigate.